src/tm2/tm-device.cpp

```diff
@@ -1061,6 +1067,49 @@ namespace librealsense
         }
     }
 
+    //mask interface
+    void tm2_sensor::set_tracking_mask(int fisheye_sensor_id, const uint8_t * mask, int width, int height, double global_ts_ms)
+    {
+        if(fisheye_sensor_id != 1 && fisheye_sensor_id != 2)
+            throw io_exception(to_string() << "Unexpected mask sensor id " << fisheye_sensor_id);
+        fisheye_sensor_id -= 1;
+        if(width != _tm_supported_profiles.mask[fisheye_sensor_id].profile.width ||
+           height != _tm_supported_profiles.mask[fisheye_sensor_id].profile.height)
+            throw io_exception(to_string() << "Unexpected tracking mask size " << width << " " << height);
+
+        perc::TrackingData::MaskFrame f;
+        f.profile.width = width;
+        f.profile.height = height;
+        f.profile.stride = width;
+        f.timestamp = global_ts_ms * 1e6; // Nanoseconds
+        f.arrivalTimeStamp = f.timestamp;
+        f.frameId = 0;
+        f.data = mask;
+        auto status = _tm_dev->SendFrame(f);
+        if (status != Status::SUCCESS)
+        {
+            LOG_WARNING("Set mask failed, status =" << (uint32_t)status);
+        }
+    }
+
+    void tm2_sensor::get_tracking_mask(int fisheye_sensor_id, uint8_t ** image, int * width, int * height, double * global_ts_ms)
+    {
+        // 1 based, as usual
+        if(fisheye_sensor_id != 1 && fisheye_sensor_id != 2)
+            throw io_exception(to_string() << "Unexpected mask sensor id " << fisheye_sensor_id);
+        int id = fisheye_sensor_id - 1;
+
+        std::lock_guard<std::mutex> lock(mask_mutex);
+        *width = mask_frames[id].profile.width;
+        *height = mask_frames[id].profile.height;
+
+        if(!mask_buffers[id]) return; // leaves image as null
+
+        int bytes = (*width) * (*height);
+        *image = (uint8_t *)malloc((*width)*(*height));
+        memcpy(*image, mask_buffers[id].get(), bytes);
+        *global_ts_ms = mask_frames[id].timestamp/1.e6; // nanoseconds to ms
```

Comment:
You mentioned that after calling `set_tracking_mask` with a timestamp the query will return nullptr till we pass that global_ts_ms .
I'm not sure, but I would expect some `if()..else()`  switch. Is implemented elsewhere ?

Reply:
Yes, look in https://github.com/IntelRealSense/librealsense/pull/5061/files#diff-3fe67583855a5e651b3830c56f877bf7R101-R106